jubatus/core/classifier/nearest_neighbor_classifier.cpp

```diff
@@ -75,13 +75,21 @@ void nearest_neighbor_classifier::train(
   }
   if (unlearner_) {
     util::concurrent::scoped_lock unlearner_lk(unlearner_mutex_);
+
+    // acquire the lock outside of touch() function
+    shared_ptr<storage::column_table> table =
+        nearest_neighbor_engine_->get_table();
+    util::concurrent::scoped_wlock lk(table->get_mutex());
+
     if (!unlearner_->touch(id)) {
       throw JUBATUS_EXCEPTION(common::exception::runtime_error(
           "cannot add new ID as number of sticky IDs reached "
           "the maximum size of unlearner: " + id));
     }
   }
-  nearest_neighbor_engine_->set_row(id, fv);
+
+  // unlearner is not called in set_row
+  nearest_neighbor_engine_->set_row(id, fv);  // lock acquired inside
```

Comment:
This acquires the write lock for the table multiple times.

http://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_rwlock_wrlock.html

> Results are undefined if the calling thread holds the read-write lock (whether a read or write lock) at the time the call is made. 

In x86, it seems there is no harm, though.